### PR TITLE
All combinations test and fixes

### PR DESCRIPTION
**Pull request by fedeallocati**
This branch adds 2 waf commands:
- **build_extensive_tests**: creates cpp files with instantiation and use of all possible combination of components for the `BOptimizer` and compiles it. The result are 503 executables. In the HAL cluster it takes ~20 minutes to do this process.
- **submit_extensive_tests**: for each of the previously created executables, it submits a job to the OAR scheduler, to run it with 2 cores and a walltime of 15 minutes(which is more than enough, none of them takes more than 5 minutes). The time to complete all the executions depends on the available resources in the cluster, but for example with just 1 free node, it took 19 minutes to finish.

Also, as a result of this extensive tests, multiple bugs where fixed:
- Exhaustive Search wasn't working.
- Changed bounds in Cmaes to get values very close to 1
- Fixed the range in Random inner opti, it was returning vectors in the range [-1, 1]
- Added unit tests for the inner optimizations
- After some talk with JB, we decided to remove MeanArchive from the mean functions.
- Fixed a bug in MaxPredictedValue, it was ignoring the dimension when passing it to the inner_opti to find the maximum gp mean value

### CODE REVIEW

File: src/tests/wscript [hunk 1]
```diff
@@ -1,41 +1,218 @@
 #!/usr/bin/env python
 # encoding: utf-8
 
+import subprocess
+import os
+
+all_combinations_init_code = """#include <boost/fusion/container/vector.hpp>
```
Comment (costashatz) [inline src/tests/wscript]: Isn't it better to put this generation of tests logic (string, func, etc) into a separate file and only use one function in the `wscript`? The `wscript` would be cleaner.

**Comment by costashatz** [comment]
I didn't let it finish in my local machine. I will push it shortly on the cluster and tell you what happens.

Nevertheless, it's not a bad idea to have this rule in the gitignore.

**Comment by fedeallocati** [comment]
No, of course it's a good idea, I just thought that it didn't worked the way it was supposed to work

**Comment by costashatz** [comment]
Compilation both on my local machine and on the cluster finished successfully (13m and 7m respectively). All tests ran smoothly also on the cluster.

File: src/tests/wscript (hunk 1 above)
Comment (jbmouret) [inline src/tests/wscript]: This template should be in its own file that python can load and fill (it would be much easier to edit).

**Comment by jbmouret** [comment]
./waf build_extensive_tests  5614.32s user 270.00s system 364% cpu 26:52.83 total

My mac is slow! I need a new one...

**Comment by fedeallocati** [comment]
Done both things

**Comment by costashatz** [comment]
I recompiled and everything seems fine.

**I approve**

**Comment by fedeallocati** [comment]
@jbmouret extracting the template to a cpp file is what you meant, right?

**Comment by jbmouret** [comment]
Yes, exactly.

On 27 Oct 2015, at 13:03, Federico Allocati <notifications@github.com> wrote:

> @jbmouret extracting the template to a cpp file is what you meant, right?
> 
> �
> Reply to this email directly or view it on GitHub.
> 

� Jean-Baptiste Mouret
http://pages.isir.upmc.fr/~mouret/

**Comment by fedeallocati** [comment]
Ok, you approve?

**Comment by jbmouret** [comment]
**OK for me**

**Pull request merged**